package com.banking.transactionservice.controller;

import com.banking.transactionservice.dto.TransactionResponse;
import com.banking.transactionservice.dto.TransferRequest;
import com.banking.transactionservice.service.TransactionService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("api/v1/transactions")
@Slf4j
@RequiredArgsConstructor
public class TransactionController {


    private final TransactionService transactionService;

    @PostMapping("/transfer")
    public ResponseEntity<TransactionResponse> transfer(
            @Valid @RequestBody TransferRequest request )
    {

        return ResponseEntity.status(HttpStatus.CREATED)
                             .body(transactionService.transfer(request));
    }

    @GetMapping("/{transactionID}")
    public ResponseEntity<TransactionResponse> getTransaction(
            @PathVariable String transactionID)
    {
        return ResponseEntity.ok(transactionService.getTransaction(transactionID));
    }

    @GetMapping("/account/{accountNumber}")
    public ResponseEntity<List<TransactionResponse>> getTransactionHistory(
            @PathVariable String accountNumber)
    {
        return ResponseEntity.ok(transactionService.getTransactionHistory(accountNumber));
    }

    @PostMapping("/{transactionID}/verify")
    public ResponseEntity<TransactionResponse> verifyOTP(
            @PathVariable String transactionID,
            @RequestParam String otp)
    {
        log.info("OTP verification request - transaction : {}",transactionID);
        return ResponseEntity.ok(
                transactionService.verifyOTP(transactionID,otp));
    }

}
